rfcs/0000-slsa-attestation.md

```diff
@@ -0,0 +1,365 @@
+      "lifecycleImage": "gcr.io/cf-build-service-public/kpack/lifecycle@sha256:0b1cd35012f7152053c42e0d6835cbb5b7c9c24207a0627f556bd931e678f8d7",
+      "rebaseImage": "gcr.io/cf-build-service-public/kpack/rebase@sha256:c0b3bbcf55750ddda69c48b1b0554470536af05556a0d639f8eb65e4a6265745",
+      "enablePriorityClasses": "false",
+      "maximumPlatformApiVersion": "",
+      "injectedSidecarSupport": "false",
+      "insecureSshTrustUnknownHosts": "true"
+    },
+    "resolvedDependencies": [
+      {
+        "uri": "git+https://github.com/buildpacks-community/kpack@refs/heads/main",
+        "digest": {
+          "sha1": "759bd290a5310bef5069385b649883e586024dae"
+        }
+      },
+      {
+        "uri": "paketobuildpacks/builder-jammy-tiny",
+        "digest": {
+          "sha256": "78d74bd1c27f633341045f1c5f7f33209f6af0a5dc5700fdfd71200b5b5a0b9a"
+        },
+        "annotations": {
+          "paketo-buildpacks/go-build": "2.1.0",
+          "paketo-buildpacks/go-mod-vendor": "1.0.25",
+          "paketo-buildpacks/node-engine": "3.0.1",
+          "paketo-buildpacks/go-dist": "2.4.2",
+          "paketo-buildpacks/yarn": "1.2.0",
+          "paketo-buildpacks/liberty": "3.8.9",
+          "paketo-buildpacks/bellsoft-liberica": "10.4.2",
+          "paketo-buildpacks/spring-boot": "5.27.5",
+          "paketo-buildpacks/google-stackdriver": "8.0.3",
+          "paketo-buildpacks/ca-certificates": "3.6.6",
+          "paketo-buildpacks/ca-certificates": "3.6.5",
+          "paketo-buildpacks/apache-tomee": "1.7.7",
+          "paketo-buildpacks/apache-tomcat": "7.13.15",
+          "paketo-buildpacks/azure-application-insights": "5.17.1",
+          "paketo-buildpacks/procfile": "5.6.7",
+          "paketo-buildpacks/procfile": "5.6.6",
+          "paketo-buildpacks/java-memory-assistant": "1.4.8",
+          "paketo-buildpacks/datadog": "4.3.0",
+          "paketo-buildpacks/encrypt-at-rest": "4.5.9",
+          "paketo-buildpacks/maven": "6.15.11",
+          "paketo-buildpacks/gradle": "7.6.1",
+          "paketo-buildpacks/sbt": "6.12.9",
+          "paketo-buildpacks/clojure-tools": "2.8.12",
+          "paketo-buildpacks/leiningen": "4.6.8",
+          "paketo-buildpacks/watchexec": "2.8.6",
+          "paketo-buildpacks/watchexec": "2.8.5",
+          "paketo-buildpacks/syft": "1.39.0",
+          "paketo-buildpacks/jattach": "1.4.8",
+          "paketo-buildpacks/executable-jar": "6.8.2",
+          "paketo-buildpacks/git": "1.0.7",
+          "paketo-buildpacks/dist-zip": "5.6.7",
+          "paketo-buildpacks/environment-variables": "4.5.6",
+          "paketo-buildpacks/environment-variables": "4.5.5",
+          "paketo-buildpacks/image-labels": "4.5.5",
+          "paketo-buildpacks/image-labels": "4.5.4",
+          "paketo-buildpacks/java": "10.3.1",
+          "paketo-buildpacks/go": "4.6.0"
+        }
```

Comment:
While this records the version of buildpacks that was used in the builder, it doesn't record the order. I feel like if you have the builder image's digest it shouldn't matter too much since you're more interested in the integrity of the build rather than reproducing it.

Reply:
Do we expect consumers of the SLSA provenance to have access to the builder?

Reply:
I'm going to interpret your question as "do we expect them to have access to the registry that the builder and app images are pushed to" and not "do we expect them to have permissions to pull the builder image". I consider permissions a human/procedural problem and not an infrastructure concern.

If they don't care about build reproducibility then it shouldn't matter, simply recording the digest is enough to verify the integrity of the build.

If they do want to be able to reproduce the build bit-for-bit, then there's really no way to do it without having access to the registry. Because if you want to completely recreate the builder from yaml, then at some point you would need the buildpackage images (and the stack images) which also requires you to have the access to the registry.

Reply:
>  I feel like if you have the builder image's digest it shouldn't matter too much since you're more interested in the integrity of the build rather than reproducing it.

How do we define the integrity of the build? I asked because this provides the buildpacks under the assumption that the builder digest and the buildpacks are sufficient. But, can't the output of the build be different with different orders of buildpacks?

Reply:
I think the core of my thinking is that anything with a checksum can be treated as a immutable basic dependency and verifying it is outside the scope of SLSA (or at least kpack's involvement in SLSA). And from my understanding, a different order would result in a different image config which would result in a different digest. 

To put it differently, I think kpack should treat the builder image as a dependency from an external system, not an artifact that was generated by kpack itself. This is to further emphasize that SLSA should apply to only the Build resource, not the entire system that kpack provides (ClusterStack, Buildpacks, Builder, Image).

Reply:
> I think the core of my thinking is that anything with a checksum can be treated as a immutable basic dependency and verifying it is outside the scope of SLSA (or at least kpack's involvement in SLSA)

The reason I hold this axiom is that it would get really ugly if we have to start considering "how can I verify that the checksums in the `go.sum` of my app is correct?"

Reply:
After thinking about it a bit more, I think I'm going to generalize the annotations section to be "the labels associated with the builder image". This would save us from doing any special parsing of the image (like figuring out buildpack version or order grouping), while surfacing more metadata (stack id, release date, distro, etc) and being more future-proof.